Source/WebGPU/WGSL/Metal/MetalCodeGenerator.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "MetalCodeGenerator.h"
+
+#include "AST.h"
+#include "MetalFunctionWriter.h"
+#include <wtf/DataLog.h>
+#include <wtf/text/StringBuilder.h>
+
+namespace WGSL {
+
+namespace Metal {
+
+static constexpr bool dumpMetalCode = false;
+
+static StringView metalCodePrologue()
+{
+    return StringView {
+        "#include <metal_stdlib>\n"
+        "\n"
+        "using namespace metal;\n"
+        "\n"_s
+    };
+
+}
+
+static void dumpMetalCodeIfNeeded(StringBuilder& stringBuilder)
+{
+    if (dumpMetalCode) {
+        dataLogLn("Generated Metal code: ");
```

Comment:
Can we use WTFLogAlways?

Reply:
Sure, I just ported this from an old implementation. Any reason to use one over the other? I don't think I've used `WTFLogAlways` before, and it wasn't obvious from the implementation.

Reply:
I heard that `dataLogLn` was a favourite of the JSC team.

Reply:
WTFLog is more fully featured; there are channels that can be toggled individually... that's what we use in WebCore.